### ISSUE TITLE
Investigation: Systematic Prediction Bias in Crypto Forecasting Models

### ISSUE THREAD

**Issue by alexoms** [OptimalMatch/temporal-trading-agents#87]
## Summary

Investigated and resolved systematic downward bias in temporal forecasting models that were predicting -10% to -40% drops for BTC regardless of market conditions when current price was $105k.

### Initial Symptoms
- All 5 ensemble models consistently predicted massive drops (-10% to -40%) over 7 days
- Predictions showed NO variation between models - all agreed on huge drops
- Pattern persisted across multiple retraining attempts
- Current BTC price: $105,021, but models predicted drops to $88k-$94k

### Investigation Process

1. **Scaler Analysis** - Examined StandardScaler statistics:
   - Mean[0] = $53,428 (historical average from 5-year dataset)
   - Scale[0] = $29,256 (price volatility)
   - Current price $105k was FAR from historical mean
   - Models were regressing predictions toward old historical average

2. **Data Period Review** - Checked training data composition:
   - Using 5-year period included 2021-2022 crash ($69k → $16k)
   - Historical mean reflected this crash period
   - Reduced to 2-year period (matching original temporal examples)
   - Problem persisted even with shorter data period

3. **Model Architecture Review** - Tested smaller models:
   - Reduced from 512-dim/6-layers to 256-dim/4-layers
   - Hypothesis: Simpler models less prone to overfit historical patterns
   - Problem still persisted

4. **Root Cause Identified** - Training on absolute prices instead of returns:
   - Models learned patterns at historical price levels ($40k-$70k range)
   - When current price was $105k, models predicted regression to historical mean
   - Predictions were NOT scale-invariant
   - Fundamental issue: absolute prices change meaning over time

## Solution: Returns-Based Training

Changed temporal forecasting to train on **percentage returns** instead of absolute prices.

### Implementation Details

**Feature Engineering Changes:**
```python
# OLD (price-based):
df['Momentum_7'] = df['Close'] - df['Close'].shift(7)  # Dollar difference
feature_columns = ['Close', 'Open', 'High', 'Low', 'Volume', ...]

# NEW (returns-based):
df['Returns'] = df['Close'].pct_change()  # Percentage change
df['Momentum_7'] = df['Returns'].rolling(7).sum()  # Cumulative return
feature_columns = ['Returns', 'Open_Returns', 'High_Returns', ...]  # Returns first!
```

**Prediction Logic Changes:**
```python
# Extract RETURNS from model forecast (not prices)
forecast_returns = forecast_np * scaler.scale_[0] + scaler.mean_[0]

# Convert returns to prices sequentially
for i, daily_return in enumerate(forecast_returns):
    cumulative_price = cumulative_price * (1 + daily_return)
    forecast_prices[i] = cumulative_price
```

**Why This Works:**
- Returns are scale-invariant: 5% move means same thing whether price is $100 or $100k
- Returns are stationary: tend to oscillate around mean of ~0%
- Model learns patterns of CHANGE, not absolute price levels
- No systematic bias toward historical price averages

### Before (Price-Based)
- **Scaler**: Mean = $53,428, Scale = $29,256
- **7-Day Predictions**: -10% to -40% drops (all models agreed)
- **Consensus**: -18.7% drop to $86k
- **Systematic bias**: Always predicted regression toward $53k historical mean

### After (Returns-Based)
- **Scaler**: Mean = -0.0016 (-0.16% daily), Scale = 0.0193 (1.93% volatility)
- **7-Day Predictions**: -1.9% to +4.6% range (realistic variation)
- **Consensus**: +0.02% (essentially flat)
- **No systematic bias**: Models show proper variation

### Individual Model Forecasts (7 days)

| Model | Price | Change | Behavior |
|-------|-------|--------|----------|
| Short-term Momentum | $103,028 | -1.9% | Slightly bearish |
| Medium-term Balanced | $105,225 | +0.2% | Nearly flat |
| Long-term Trend | $106,226 | +1.1% | Slightly bullish |
| Mean Reversion | $109,834 | +4.6% | Most bullish |
| Mid-term Momentum | $106,147 | +1.1% | Slightly bullish |

Models now show **realistic variation** and proper forecasting behavior!

### Files Changed
- `examples/crypto_ensemble_forecast.py` - Returns-based feature engineering and prediction
- `backend/main.py` - Data period selection
- `backend/requirements.txt` - Dependencies
- `strategies/strategy_utils.py` - Strategy utilities

### Key Code Changes
1. Calculate Returns for all OHLCV data (`pct_change()`)
2. Make 'Returns' PRIMARY feature at index 0 (what model predicts)
3. Convert momentum from price diffs to cumulative returns
4. Update prediction logic to convert returns → prices
5. Fit scaler on recent 90-day window (not full history)

### Validation
- ✅ All 5 models train successfully
- ✅ Scaler shows returns-based stats (mean ≈ 0%, scale ≈ 2%)
- ✅ Predictions in realistic range (-2% to +5%)
- ✅ No systematic bias
- ✅ Models show proper variation
- ✅ Consensus forecast reasonable (+0.02%)

## Lessons Learned

1. **Scale Invariance Matters**: Financial models MUST work across different price regimes
2. **Returns vs Prices**: Training on returns eliminates scale-dependency
3. **Historical Bias**: Using absolute prices creates bias toward historical averages
4. **Feature Selection**: Primary prediction feature determines model behavior
5. **Stationarity**: Returns are more stationary than prices (better for ML)

## References

- PR #86: Implementation of returns-based training
- Original temporal forecasting examples use similar approach
- Standard practice in quantitative finance

## Impact

This fix enables realistic cryptocurrency price forecasting that:
- Works across different price regimes ($10k, $100k, $1M)
- Shows proper model variation and disagreement
- Eliminates systematic bias toward historical averages
- Makes predictions interpretable and actionable

The fundamental change from price-based to returns-based training solved a critical bug that was making the forecasting system unusable.
